File: server/src/modules/heatmapParser.ts
import Papa from 'papaparse';
import type { RawHeatmapRow, HeatmapRow, ValidationResult, ValidationError } from '../types.js';

const REQUIRED_COLUMNS = ['Date', 'Program', 'Interval_Start_Time', 'Over_Under_Value'] as const;
// Lobby is optional — defaults to empty string if missing

/**
 * Normalizes a date string to YYYY-MM-DD format.
 * Accepts: "4/5/2026", "04/05/2026", "2026-04-05", etc.
 */
export function normalizeDate(dateStr: string): string | null {
  const trimmed = dateStr.trim();
  if (/^\d{4}-\d{2}-\d{2}$/.test(trimmed)) return trimmed;

  const slashMatch = /^(\d{1,2})\/(\d{1,2})\/(\d{4})$/.exec(trimmed);
  if (slashMatch) {
    const m = slashMatch[1].padStart(2, '0');
    const d = slashMatch[2].padStart(2, '0');
    return `${slashMatch[3]}-${m}-${d}`;
  }

  const months: Record<string, string> = {
    jan: '01', feb: '02', mar: '03', apr: '04', may: '05', jun: '06',
    jul: '07', aug: '08', sep: '09', oct: '10', nov: '11', dec: '12',
  };
  const dMonMatch = /^(\d{1,2})-([A-Za-z]{3})-(\d{4})$/.exec(trimmed);
  if (dMonMatch) {
    const mon = months[dMonMatch[2].toLowerCase()];
    if (mon) {
      return `${dMonMatch[3]}-${mon}-${dMonMatch[1].padStart(2, '0')}`;
    }
  }

  return null;
}

/**
 * Normalizes a time string to HH:MM format (24-hour).
 */
export function normalizeInterval(time: string): string | null {
  const trimmed = time.trim().toUpperCase();

  const match24 = /^(\d{1,2}):(\d{1,2})$/.exec(trimmed);
  if (match24) {
    const h = parseInt(match24[1], 10);
    const m = parseInt(match24[2], 10);
    if (h >= 0 && h <= 23 && (m === 0 || m === 30)) {
      return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}`;
    }
    return null;
  }

  const match12 = /^(\d{1,2}):(\d{2})\s*(AM|PM)$/.exec(trimmed);
  if (match12) {
    let h = parseInt(match12[1], 10);
    const m = parseInt(match12[2], 10);
    const period = match12[3];
    if (h < 1 || h > 12 || (m !== 0 && m !== 30)) return null;
    if (period === 'AM' && h === 12) h = 0;
    if (period === 'PM' && h !== 12) h += 12;
    return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}`;
  }

  return null;
}

/**
 * Validates that a time string is a valid half-hour interval.
 */
export function isValidHalfHourInterval(time: string): boolean {
  return normalizeInterval(time) !== null;
}

/**
 * Validates an array of raw heatmap rows.
 */
export function validateHeatmapRows(rows: RawHeatmapRow[]): ValidationResult<HeatmapRow> {
  const valid: HeatmapRow[] = [];
  const errors: ValidationError[] = [];

  if (rows.length > 0) {
    const columns = Object.keys(rows[0]);
    for (const col of REQUIRED_COLUMNS) {
      if (!columns.includes(col)) {
        errors.push({ row: 0, field: col, message: `Missing required column: ${col}` });
      }
    }
    if (errors.length > 0) return { valid, errors };
  }

  for (let i = 0; i < rows.length; i++) {
    const row = rows[i];
    const rowNum = i + 1;
    let rowValid = true;

    if (!row.Date || row.Date.trim() === '') {
      errors.push({ row: rowNum, field: 'Date', message: `Row ${rowNum}: Date is required` });
      rowValid = false;
    } else {
      const normalized = normalizeDate(row.Date.trim());
      if (!normalized) {
        errors.push({ row: rowNum, field: 'Date', message: `Row ${rowNum}: Date format not recognized` });
        rowValid = false;
      }
    }

    if (!row.Program || row.Program.trim() === '') {
      errors.push({ row: rowNum, field: 'Program', message: `Row ${rowNum}: Program is required` });
      rowValid = false;
    }

    if (!row.Interval_Start_Time || !isValidHalfHourInterval(row.Interval_Start_Time.trim())) {
      errors.push({
        row: rowNum,
        field: 'Interval_Start_Time',
        message: `Row ${rowNum}: Interval_Start_Time must be a valid half-hour interval`,
      });
      rowValid = false;
    }

    const overUnderStr = row.Over_Under_Value?.trim() ?? '';
    if (overUnderStr === '' || isNaN(Number(overUnderStr))) {
      errors.push({
        row: rowNum,
        field: 'Over_Under_Value',
        message: `Row ${rowNum}: Over_Under_Value must be numeric`,
      });
      rowValid = false;
    }

    if (rowValid) {
      valid.push({
        date: normalizeDate(row.Date!.trim())!,
        program: row.Program!.trim(),
        lobby: row.Lobby?.trim() ?? '',
        intervalStartTime: normalizeInterval(row.Interval_Start_Time!.trim())!,
        overUnderValue: Number(overUnderStr),
      });
    }
  }

  return { valid, errors };
}

/**
 * Known month abbreviations for pivot date column headers like "31-May", "1-Jun".
 */
const MONTH_ABBR: Record<string, string> = {
  jan: '01', feb: '02', mar: '03', apr: '04', may: '05', jun: '06',
  jul: '07', aug: '08', sep: '09', oct: '10', nov: '11', dec: '12',
};

/**
 * Checks if a column header looks like a date (e.g. "31-May", "1-Jun", "15-Dec").
 */
function isDateColumn(header: string): boolean {
  const match = /^(\d{1,2})-([A-Za-z]{3})$/.exec(header.trim());
  if (!match) return false;
  return MONTH_ABBR[match[2].toLowerCase()] !== undefined;
}

/**
 * Converts a pivot date column header (e.g. "31-May") to YYYY-MM-DD.
 * Uses the current year by default; if the resulting date is more than
 * 6 months in the past, assumes next year.
 */
export function pivotDateToISO(header: string): string | null {
  const match = /^(\d{1,2})-([A-Za-z]{3})$/.exec(header.trim());
  if (!match) return null;
  const day = match[1].padStart(2, '0');
  const mon = MONTH_ABBR[match[2].toLowerCase()];
  if (!mon) return null;

  const now = new Date();
  let year = now.getFullYear();
  const candidate = new Date(`${year}-${mon}-${day}T00:00:00`);
  const sixMonthsAgo = new Date(now);
  sixMonthsAgo.setMonth(sixMonthsAgo.getMonth() - 6);
  if (candidate < sixMonthsAgo) {
    year += 1;
  }
  return `${year}-${mon}-${day}`;
}

/**
 * Converts a pivot-style interval (e.g. "0000", "0030", "1430") to HH:MM.
 * Also accepts HH:MM format directly (e.g. "07:00", "7:00").
 */
export function pivotIntervalToTime(interval: string): string | null {
  const trimmed = interval.trim();

  // Try HH:MM or H:MM format first (from normalizeInterval)
  const normalized = normalizeInterval(trimmed);
  if (normalized) return normalized;

  // Try 4-digit format (e.g. "0000", "0030", "1430")
  const match = /^(\d{2})(\d{2})$/.exec(trimmed);
  if (!match) return null;
  const h = parseInt(match[1], 10);
  const m = parseInt(match[2], 10);
  if (h < 0 || h > 23 || (m !== 0 && m !== 30)) return null;
  return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}`;
}

/**
 * Detects whether parsed CSV data is in pivot format.
 * Pivot format has an "Interval" column and date columns like "31-May".
 */
export function isPivotFormat(headers: string[]): boolean {
  const hasInterval = headers.some(h => h.trim().toLowerCase() === 'interval');
  const dateColumns = headers.filter(isDateColumn);
  return hasInterval && dateColumns.length > 0;
}

/**
 * Converts pivot-format rows into standard RawHeatmapRow format.
 * Pivot format: Program, Lobby (optional), Interval, 31-May, 1-Jun, ...
 * Standard format: Date, Program, Lobby, Interval_Start_Time, Over_Under_Value
 */
export function convertPivotToStandard(
  headers: string[],
  rows: Record<string, string>[],
): RawHeatmapRow[] {
  const dateColumns = headers.filter(isDateColumn);
  const hasProgram = headers.some(h => h.trim().toLowerCase() === 'program');
  const hasLobby = headers.some(h => h.trim().toLowerCase() === 'lobby');
  const intervalKey = headers.find(h => h.trim().toLowerCase() === 'interval') ?? 'Interval';
  const programKey = headers.find(h => h.trim().toLowerCase() === 'program') ?? 'Program';
  const lobbyKey = headers.find(h => h.trim().toLowerCase() === 'lobby') ?? 'Lobby';

  const result: RawHeatmapRow[] = [];

  for (const row of rows) {
    const rawInterval = row[intervalKey]?.trim() ?? '';
    const program = hasProgram ? (row[programKey]?.trim() ?? '') : '';
    const lobby = hasLobby ? (row[lobbyKey]?.trim() ?? '') : '';

    for (const dateCol of dateColumns) {
      const value = row[dateCol]?.trim() ?? '';
      if (value === '') continue;

      const isoDate = pivotDateToISO(dateCol);
      const time = pivotIntervalToTime(rawInterval);

      result.push({
        Date: isoDate ?? dateCol,
        Program: program,
        Lobby: lobby,
        Interval_Start_Time: time ?? rawInterval,
        Over_Under_Value: value,
      });
    }
  }

  return result;
}

/**
 * Parses a heatmap CSV string and validates the rows.
 * Auto-detects pivot format (Interval × Date columns) and converts it
 * to the standard vertical format before validation.
 */
export function parseHeatmapCSV(csvString: string): ValidationResult<HeatmapRow> {
  const results = Papa.parse<Record<string, string>>(csvString, {
    header: true,
    skipEmptyLines: true,
  });

  const headers = results.meta.fields ?? [];

  if (isPivotFormat(headers)) {
    const converted = convertPivotToStandard(headers, results.data);
    return validateHeatmapRows(converted);
  }

  return validateHeatmapRows(results.data as RawHeatmapRow[]);
}

/**
 * Parses a D-Mon or DD-Mon date header (e.g., "31-May", "1-Jun") to YYYY-MM-DD.
 * Infers year from referenceYear (defaults to current year).
 * Handles Dec→Jan rollover: if the parsed month is more than 6 months before
 * the reference month, assume next year.
 */
export function parseDateHeader(header: string, referenceYear?: number): string | null {
  const months: Record<string, number> = {
    jan: 1, feb: 2, mar: 3, apr: 4, may: 5, jun: 6,
    jul: 7, aug: 8, sep: 9, oct: 10, nov: 11, dec: 12,
  };
  const match = /^(\d{1,2})-([A-Za-z]{3})$/.exec(header.trim());
  if (!match) return null;
  const day = parseInt(match[1], 10);
  const mon = months[match[2].toLowerCase()];
  if (!mon) return null;
  if (day < 1 || day > 31) return null;

  const now = new Date();
  let year = referenceYear ?? now.getFullYear();

  // Handle Dec→Jan rollover: if the header month is Jan-Feb and current month is Nov-Dec,
  // assume next year
  const refMonth = now.getMonth() + 1; // 1-based
  if (mon <= 2 && refMonth >= 11 && referenceYear === undefined) {
    year += 1;
  }

  return `${year}-${String(mon).padStart(2, '0')}-${String(day).padStart(2, '0')}`;
}

/**
 * Detects whether the CSV columns represent a pivot/crosstab format.
 * Pivot format has Program + Interval columns but lacks Date and Over_Under_Value,
 * and has date-like column headers (D-Mon pattern).
 */
export function detectPivotFormat(columns: string[]): { isPivot: boolean; dateColumns: string[] } {
  const hasInterval = columns.some(c => c.trim() === 'Interval');
  const hasDate = columns.some(c => c.trim() === 'Date');
  const hasOverUnder = columns.some(c => c.trim() === 'Over_Under_Value');

  if (!hasInterval || hasDate || hasOverUnder) {
    return { isPivot: false, dateColumns: [] };
  }

  const dateColumns: string[] = [];
  for (const col of columns) {
    if (parseDateHeader(col.trim()) !== null) {
      dateColumns.push(col);
    }
  }

  return {
    isPivot: dateColumns.length > 0,
    dateColumns,
  };
}

/**
 * Transposes pivot-format rows into flat RawHeatmapRow records.
 * For each row × each date column, produces one RawHeatmapRow with:
 *   Date = parseDateHeader(dateColumnHeader)
 *   Program = row.Program
 *   Interval_Start_Time = row.Interval
 *   Over_Under_Value = cell value
 * Skips cells that are empty or whitespace-only.
 */
export function transposePivotRows(
  rawRows: Record<string, string>[],
  dateColumns: string[],
  refYear?: number,
): RawHeatmapRow[] {
  const result: RawHeatmapRow[] = [];

  for (const row of rawRows) {
    const program = row['Program']?.trim() ?? '';
    const lobby = row['Lobby']?.trim() ?? '';
    const interval = row['Interval']?.trim() ?? '';

    for (const dateCol of dateColumns) {
      const cellValue = (row[dateCol] ?? '').trim();
      if (cellValue === '') continue;

      const parsedDate = parseDateHeader(dateCol.trim(), refYear);
      if (!parsedDate) continue;

      result.push({
        Date: parsedDate,
        Program: program,
        Lobby: lobby,
        Interval_Start_Time: interval,
        Over_Under_Value: cellValue,
      });
    }
  }

  return result;
}

File: python-server/modules/heatmap_parser.py
"""
Heatmap CSV parser with pivot format auto-detection.
Mirrors heatmapParser.ts exactly.
"""
import csv
import io
import re
from datetime import date, datetime

REQUIRED_COLUMNS = ['Date', 'Program', 'Interval_Start_Time', 'Over_Under_Value']

MONTH_ABBR = {
    'jan': '01', 'feb': '02', 'mar': '03', 'apr': '04',
    'may': '05', 'jun': '06', 'jul': '07', 'aug': '08',
    'sep': '09', 'oct': '10', 'nov': '11', 'dec': '12',
}


def normalize_date(date_str):
    """Normalizes a date string to YYYY-MM-DD format."""
    trimmed = date_str.strip()
    if re.match(r'^\d{4}-\d{2}-\d{2}$', trimmed):
        return trimmed

    m = re.match(r'^(\d{1,2})/(\d{1,2})/(\d{4})$', trimmed)
    if m:
        month = m.group(1).zfill(2)
        day = m.group(2).zfill(2)
        return f"{m.group(3)}-{month}-{day}"

    m = re.match(r'^(\d{1,2})-([A-Za-z]{3})-(\d{4})$', trimmed)
    if m:
        mon = MONTH_ABBR.get(m.group(2).lower())
        if mon:
            return f"{m.group(3)}-{mon}-{m.group(1).zfill(2)}"

    return None


def normalize_interval(time_str):
    """Normalizes a time string to HH:MM format (24-hour)."""
    trimmed = time_str.strip().upper()

    m24 = re.match(r'^(\d{1,2}):(\d{1,2})$', trimmed)
    if m24:
        h = int(m24.group(1))
        mins = int(m24.group(2))
        if 0 <= h <= 23 and mins in (0, 30):
            return f"{str(h).zfill(2)}:{str(mins).zfill(2)}"
        return None

    m12 = re.match(r'^(\d{1,2}):(\d{2})\s*(AM|PM)$', trimmed)
    if m12:
        h = int(m12.group(1))
        mins = int(m12.group(2))
        period = m12.group(3)
        if h < 1 or h > 12 or mins not in (0, 30):
            return None
        if period == 'AM' and h == 12:
            h = 0
        if period == 'PM' and h != 12:
            h += 12
        return f"{str(h).zfill(2)}:{str(mins).zfill(2)}"

    return None


def is_valid_half_hour_interval(time_str):
    return normalize_interval(time_str) is not None


def _is_date_column(header):
    """Checks if a column header looks like a date (e.g. '31-May', '1-Jun')."""
    m = re.match(r'^(\d{1,2})-([A-Za-z]{3})$', header.strip())
    if not m:
        return False
    return m.group(2).lower() in MONTH_ABBR


def pivot_date_to_iso(header):
    """Converts a pivot date column header (e.g. '31-May') to YYYY-MM-DD."""
    m = re.match(r'^(\d{1,2})-([A-Za-z]{3})$', header.strip())
    if not m:
        return None
    day = m.group(1).zfill(2)
    mon = MONTH_ABBR.get(m.group(2).lower())
    if not mon:
        return None

    now = datetime.now()
    year = now.year
    try:
        candidate = datetime(year, int(mon), int(day))
    except ValueError:
        return None
    from dateutil.relativedelta import relativedelta
    six_months_ago = now - relativedelta(months=6)
    if candidate < six_months_ago:
        year += 1
    return f"{year}-{mon}-{day}"


def _pivot_date_to_iso_simple(header):
    """Converts pivot date header without dateutil dependency."""
    m = re.match(r'^(\d{1,2})-([A-Za-z]{3})$', header.strip())
    if not m:
        return None
    day = m.group(1).zfill(2)
    mon = MONTH_ABBR.get(m.group(2).lower())
    if not mon:
        return None

    now = datetime.now()
    year = now.year
    try:
        candidate = datetime(year, int(mon), int(day))
    except ValueError:
        return None
    # 6 months ago approximation
    six_months_ago_month = now.month - 6
    six_months_ago_year = now.year
    if six_months_ago_month <= 0:
        six_months_ago_month += 12
        six_months_ago_year -= 1
    try:
        six_months_ago = datetime(six_months_ago_year, six_months_ago_month, min(now.day, 28))
    except ValueError:
        six_months_ago = datetime(six_months_ago_year, six_months_ago_month, 28)
    if candidate < six_months_ago:
        year += 1
    return f"{year}-{mon}-{day}"


def pivot_interval_to_time(interval):
    """Converts a pivot-style interval (e.g. '0000', '0030', '1430', '0:0', '1:30') to HH:MM.
    Tries normalize_interval first to handle formats like '0:0', '1:30', then falls back
    to the 4-digit format."""
    trimmed = interval.strip()

    # Try normalize_interval first (handles "0:0", "1:30", "07:00", etc.)
    normalized = normalize_interval(trimmed)
    if normalized is not None:
        return normalized

    # Fall back to 4-digit format (e.g. '0000', '0030', '1430')
    m = re.match(r'^(\d{2})(\d{2})$', trimmed)
    if not m:
        return None
    h = int(m.group(1))
    mins = int(m.group(2))
    if h < 0 or h > 23 or mins not in (0, 30):
        return None
    return f"{str(h).zfill(2)}:{str(mins).zfill(2)}"


def is_pivot_format(headers):
    """Detects whether parsed CSV data is in pivot format."""
    has_interval = any(h.strip().lower() == 'interval' for h in headers)
    date_columns = [h for h in headers if _is_date_column(h)]
    return has_interval and len(date_columns) > 0


def convert_pivot_to_standard(headers, rows):
    """Converts pivot-format rows into standard RawHeatmapRow format."""
    date_columns = [h for h in headers if _is_date_column(h)]
    interval_key = next((h for h in headers if h.strip().lower() == 'interval'), 'Interval')
    program_key = next((h for h in headers if h.strip().lower() == 'program'), 'Program')
    lobby_key = next((h for h in headers if h.strip().lower() == 'lobby'), 'Lobby')
    has_program = any(h.strip().lower() == 'program' for h in headers)
    has_lobby = any(h.strip().lower() == 'lobby' for h in headers)

    result = []
    for row in rows:
        raw_interval = (row.get(interval_key) or '').strip()
        program = (row.get(program_key) or '').strip() if has_program else ''
        lobby = (row.get(lobby_key) or '').strip() if has_lobby else ''

        for date_col in date_columns:
            value = (row.get(date_col) or '').strip()
            if not value:
                continue
            iso_date = _pivot_date_to_iso_simple(date_col)
            time_val = pivot_interval_to_time(raw_interval)
            result.append({
                'Date': iso_date or date_col,
                'Program': program,
                'Lobby': lobby,
                'Interval_Start_Time': time_val or raw_interval,
                'Over_Under_Value': value,
            })
    return result


def validate_heatmap_rows(rows):
    """Validates an array of raw heatmap rows."""
    valid = []
    errors = []

    if rows:
        columns = list(rows[0].keys())
        for col in REQUIRED_COLUMNS:
            if col not in columns:
                errors.append({'row': 0, 'field': col, 'message': f'Missing required column: {col}'})
        if errors:
            return {'valid': valid, 'errors': errors}

    for i, row in enumerate(rows):
        row_num = i + 1
        row_valid = True

        date_val = (row.get('Date') or '').strip()
        if not date_val:
            errors.append({'row': row_num, 'field': 'Date', 'message': f'Row {row_num}: Date is required'})
            row_valid = False
        else:
            normalized_date = normalize_date(date_val)
            if not normalized_date:
                errors.append({'row': row_num, 'field': 'Date', 'message': f'Row {row_num}: Date format not recognized'})
                row_valid = False

        program_val = (row.get('Program') or '').strip()
        if not program_val:
            errors.append({'row': row_num, 'field': 'Program', 'message': f'Row {row_num}: Program is required'})
            row_valid = False

        interval_val = (row.get('Interval_Start_Time') or '').strip()
        if not interval_val or not is_valid_half_hour_interval(interval_val):
            errors.append({
                'row': row_num, 'field': 'Interval_Start_Time',
                'message': f'Row {row_num}: Interval_Start_Time must be a valid half-hour interval',
            })
            row_valid = False

        over_under_str = (row.get('Over_Under_Value') or '').strip()
        if not over_under_str:
            errors.append({
                'row': row_num, 'field': 'Over_Under_Value',
                'message': f'Row {row_num}: Over_Under_Value must be numeric',
            })
            row_valid = False
        else:
            try:
                float(over_under_str)
            except ValueError:
                errors.append({
                    'row': row_num, 'field': 'Over_Under_Value',
                    'message': f'Row {row_num}: Over_Under_Value must be numeric',
                })
                row_valid = False

        if row_valid:
            valid.append({
                'date': normalize_date(date_val),
                'program': program_val,
                'lobby': (row.get('Lobby') or '').strip(),
                'intervalStartTime': normalize_interval(interval_val),
                'overUnderValue': float(over_under_str),
            })

    return {'valid': valid, 'errors': errors}


def parse_heatmap_csv(csv_string):
    """Parses a heatmap CSV string and validates the rows.
    Auto-detects pivot format and converts before validation."""
    reader = csv.DictReader(io.StringIO(csv_string))
    headers = reader.fieldnames or []
    rows = list(reader)

    if is_pivot_format(headers):
        converted = convert_pivot_to_standard(headers, rows)
        return validate_heatmap_rows(converted)

    return validate_heatmap_rows(rows)
